### ISSUE TITLE
Replies are sometimes displayed out of order

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop-client#653]
# STR 1 - For `SendReplyJobTimeoutError`

1. Apply this diff so that sometimes a reply fails:

```diff
--- a/securedrop_client/api_jobs/uploads.py
+++ b/securedrop_client/api_jobs/uploads.py
@@ -28,6 +28,11 @@ class SendReplyJob(ApiJob):
         database and return the reply uuid string. Otherwise raise a SendReplyJobException so that
         we can return the reply uuid.
         '''
+        import random
+        if random.choice([0, 1]):
+            self.remaining_attempts = 0
+            raise SendReplyJobTimeoutError('test', self.reply_uuid)
+
         try:
             draft_reply_db_object = session.query(DraftReply).filter_by(uuid=self.reply_uuid).one()
             source = session.query(Source).filter_by(uuid=self.source_uuid).one()
```
2. Send many replies so that there are some successful and once one fails see the conversation reorder like so:

![Screenshot from 2020-02-11 11-40-00](https://user-images.githubusercontent.com/4522213/74272410-6a102c80-4cc3-11ea-881c-ea5583cc8927.png)

3. Wait until the reply succeeds (because of auto-resume when sync succeeds) and see that the failed reply moves to a successful state

![Screenshot from 2020-02-11 11-40-13](https://user-images.githubusercontent.com/4522213/74272467-8318dd80-4cc3-11ea-8932-5625ce70d1b4.png)

4. Wait until the next sync and see the replies get reordered correctly:

![Screenshot from 2020-02-11 11-40-25](https://user-images.githubusercontent.com/4522213/74272482-8f9d3600-4cc3-11ea-90c9-15c102ea9a94.png)

# STR 2 - For `SendReplyJobError`

1. Apply this diff so that sometimes a reply fails:

```diff
--- a/securedrop_client/api_jobs/uploads.py
+++ b/securedrop_client/api_jobs/uploads.py
@@ -28,6 +28,11 @@ class SendReplyJob(ApiJob):
         database and return the reply uuid string. Otherwise raise a SendReplyJobException so that
         we can return the reply uuid.
         '''
+        import random
+        if random.choice([0, 1]):
+            self.remaining_attempts = 0
+            raise SendReplyJobError('test', self.reply_uuid)
+
         try:
             draft_reply_db_object = session.query(DraftReply).filter_by(uuid=self.reply_uuid).one()
             source = session.query(Source).filter_by(uuid=self.source_uuid).one()
```
repeat steps 2-4 above

**Comment by redshiftzero** [comment]
Problem 1: Replies are prioritized low in the queue

Fixes: 
 * #462 (we're clogging up the queue with Metadata jobs)
 * we prioritize the user-triggered actions higher - https://github.com/freedomofpress/securedrop-client/compare/deprioritize-downloads
* eventually: allow jobs that can run concurrently to do so

Problem 2: conversation only updates

Fix:
* As you notice these replies will reorder themselves, but the conversation view might not update immediately. We could have forced the entire conversation view to be redrawn every time a reply is sent but did not due to #473 which needs to be done (we are deleting the _entire_ view and redrawing it but only when a sync completes (if we do it more often we see flicker which is also a bug), which is why you see the reply reorder itself into the right position after a little while).

**Comment by eloquence** [comment]
>  #462 (we're clogging up the queue with Metadata jobs)

Agreed, but I'm not sure that this is the entire cause of the reply slowness I'm seeing -- I was able to reliably reproduce this over the course of a couple of minutes, where it would alternate between fast (2.5 seconds) and slow (15-20 seconds), without me initiating any sync of my own during this time.

**Comment by eloquence** [comment]
Performance is significantly better in the latest nightly build, I'm now seeing >1st replies transition all the way to "sent" in the 6-8 second range. Perhaps this will go down further once all the metadata sync changes have landed.

We still have the ordering problem; the order is only "eventually correct", but initially replies appear out of order. 

Also worth noting: Out of 20 replies or so I encountered one random failure. I would have expected it to go to "Failed to send" after lots more retries, but it basically transitioned to "Failed" almost immediately.

**Comment by ntoll** [comment]
Is this not fixed by #688..?

**Comment by redshiftzero** [comment]
yep you're right, I have not observed this behavior for some time. Closing and let's reopen if we see this again.

**Comment by creviera** [comment]
I just updated the description with the latest STR and reopening because this is not resolved yet.

**Comment by ntoll** [comment]
Aha..! :smile:

**Comment by eloquence** [comment]
I've not tested your STR yet Allie, but yay, generally speaking, replies _are_ a lot more reliable now (in Qubes). I'm not seeing the ordering issue during normal use, and replies are sent quite quickly.

**Comment by creviera** [comment]
When someone gets a chance, could you run through the updated STR to confirm that this is a bug we need to fix: https://github.com/freedomofpress/securedrop-client/issues/653#issue-535526020 -- I'm not a 100% sure that @ntoll has already done so

**Comment by creviera** [comment]
today, when running through the STR I saw the same issue: 
![Screenshot from 2020-02-24 17-05-22](https://user-images.githubusercontent.com/4522213/75205188-f55bda00-5727-11ea-9e9d-1947add9cfae.png)

**Comment by creviera** [comment]
Note: these replies stay out of order even after a pending reply turns to a successful reply. when i close the client and reopen the replies are reordered correctly:

![Screenshot from 2020-02-24 17-10-01](https://user-images.githubusercontent.com/4522213/75205452-a82c3800-5728-11ea-96e1-89eb151f2302.png)

It appears the bug is somewhere within [update_conversation](https://github.com/freedomofpress/securedrop-client/blob/8adf1f25226db8cf7232054953808418d51ed527/securedrop_client/gui/widgets.py#L2729-L2747)

**Comment by ntoll** [comment]
So... @creviera yes I did manage to recreate but functional tests have taken up my time. However, now they're done I'll look into this.

I've just observed the following (problem) behaviour by following the STR:

* A failed message appears to block the queue of pending messages (i.e. if message 1 is failed, the subsequent messages 2, 3, and 4 all appear in the greyed out pending state). This appears to suggest @redshiftzero is correct about queue priorities / concurrent replies.
* Once the client retries after 15 seconds the failed message is marked as sent and appears in the *correct* position in the queue -- however, the pending and any subsequent failed messages are still in the wrong place. For instance if message 1 had succeeded, message 2 failed and messages 3 and 4 were pending, then, when message 2 succeeded it would appear immediately after message 1 but messages 3 and 4 might be in the wrong place.

I agree with @creviera that the error is in 'update_conversation' and that it's not so much a bug, as an omission to deal with how to order pending or failed replies (which won't yet be known by the server, so the message order when we next refresh **will be correct** but won't contain those unknown replies).

My suggested fix (which I'll go ahead and implement so we can at least confirm my hunch is correct -- we can change the fix's behaviour during review of the impending PR) is:

* All replies known to the server retain their position in the order of the conversation.
* All replies unknown to the server (i.e. those in failed or pending states) **MUST** always be the last replies in the conversation list.

Sound like a plan..?

Give me 10 mins to get a PR in fight with a potential fix / unit tests to prove things behave as expected.

**Comment by redshiftzero** [comment]
OK having trouble reproducing this due to #820 but looking at the logic: we [have logic](https://github.com/freedomofpress/securedrop-client/blob/b75db2d9d937060d87dfe54755615f89734b2b0a/securedrop_client/storage.py#L396) to reorder drafts to ensure consistent ordering of interleaved successful replies and drafts. This is performed at sync time, and at reply send time when the reply a successful. That means that the local order we have stored should be correct. In order for this order to be shown in the UI, we need to enter `update_conversation` for the reordering logic to run at the UI level. Is that happening? I'm not seeing where if so - it seems like only on source click and on recreation of the conversation view. That would explain the behavior.

Question: do we really need [`add_reply_from_reply_box`](https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/gui/widgets.py#L2843)? This is one path to widgets getting added to the conversation view outside of calls to `update_conversation`. It seems unnecessary now: in `on_reply_sent` [here](https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/gui/widgets.py#L2865) we should be able to call `update_conversation` instead of this additional method `add_reply_from_reply_box` (`source.collection` has the `DraftReply` object which will be created immediately when we click send).

**Comment by creviera** [comment]
[add_reply_from_reply_box](https://github.com/freedomofpress/securedrop-client/blob/master/securedrop_client/gui/widgets.py#L2843) is called when the user clicks "Send". Without it, we won't immediately see replies in the ConversationView. If we remove it, we will need to call `update_conversation` so that [this code will add the new draft reply](https://github.com/freedomofpress/securedrop-client/blob/b75db2d9d937060d87dfe54755615f89734b2b0a/securedrop_client/gui/widgets.py#L2779-L2780). I think this means we will need a new signal to emit from the controller [after it creates the new DraftReply](https://github.com/freedomofpress/securedrop-client/blob/b75db2d9d937060d87dfe54755615f89734b2b0a/securedrop_client/logic.py#L773-L803) in response to the user clicking "Send." 

I think this is a good refactor, but I'm not sure it's related to this issue, because `update_conversation` uses an index to reorder conversation items if the index no longer matches its GUI placement in the conversation, see https://github.com/freedomofpress/securedrop-client/blob/b75db2d9d937060d87dfe54755615f89734b2b0a/securedrop_client/gui/widgets.py#L2758-L2768

The strange reordering behavior only seems to be introduced once a reply is failed, which is why I no longer see STR #1 reported in this issue after making this change: https://github.com/freedomofpress/securedrop-client/pull/818

**Comment by redshiftzero** [comment]
In order to verify if this was a UI only issue or not I decided to double check using the debug logging [here](https://github.com/freedomofpress/securedrop-client/tree/logging-niceness) along with logging the current value of `source.conversation` when drawing the conversation view. I noticed that the `interaction_count` was not getting updated, which led me to #829, from the comment:

>   We use the file_counter on each Reply, Message, File, and DraftReply object to order the conversation view. We expect a unique file_counter for Reply, Message, and File objects since they are retrieved from the server. For DraftReply, we don't have that unique constraint, and we instead expect the file_counter to be the interaction_count at the time of send. If we do not update the interaction_count after each successful reply send, future drafts will have an interaction_count that is too low, leading to incorrectly ordered drafts until a metadata sync completes (the metadata sync is the place where the source object is updated from the server).

**Comment by rocodes** [comment]
I've noticed a sort-of a recurrence of this issue, but no failed reply is required. (Unreliable STR: send lots of replies in quick succession, can observe mis-ordered messages some of the time, order corrects after sync). I'm not sure if it merits re-opening this issue, since it basically involves quickly sending many consecutive replies, and is resolved after a sync, but documenting it here.)
 
(sometimes the mis-ordering is worse than this, sometimes it doesn't occur at all even with upwards of 15 messages):
![order](https://user-images.githubusercontent.com/22777700/157745427-dd83b7f3-696b-4dce-969c-37c9da2675bd.png)

(reordered correctly after a sync)
![order2](https://user-images.githubusercontent.com/22777700/157745425-81b88cce-cb44-4f91-9f1a-becc3ef13a1a.png)

**Comment by eloquence** [comment]
I saw this once as well during recent QA but didn't write it up, thanks for noting it here. I'd say let's definitely reopen, defer to Allie on priority, but I think it'd be good to know at least what's causing it.

**Comment by eloquence** [comment]
Closing as an edge case for the old client; should be considered as part of our new client test suite (made a note of that and will incorporate it into issues-to-be-created for test coverage).
